### DISCUSSION TITLE
Volume not found errors with EnableCacheStreamedVolumes

### DISCUSSION THREAD

**Discussion by andy-paine** [concourse/concourse#8083] [General]
## Background
Enabling the cache streamed volumes (together with #8047) seems to allow us to scale up workers when the task queue increases and quickly/efficiently seed the required volumes into them. Without this feature, we will not be able to dynamically scale based on the task queue as the tasks only run/queue up after the `get` steps have run. This results in the `get` resource volumes never being cached on new workers and thus require streaming for every step scheduled on them unless we use `EnableCacheStreamedVolumes`.

## Issue
We have been running into an issue where during streaming of volumes and files (`metadata.json`, `task.yml`), we are getting volume-not-found errors from the worker with the volume/files to be streamed. These not-found errors come after successful streams and any streams started before the first error complete successfully (even if the volume has been “gone” for multiple minutes). Turning off any scaling appears to have resolved this issue so far (as of 15/02) but this is obviously not practical as the eventual goal is for dynamic scaling. Whilst it appears to have resolved the issue, it also does not correlate that well with the behaviour we are seeing as the instance that loses the volume is not scaled in at any point during the process. Looking through debug level logs for both the web and worker process, it does not appear that Concourse itself is deleting the volume as no `destroy-volume` logs are emitted for the volume in question. Volume sweeping and destruction is taking place (as visible in the logs) but never targets the volume in question at the time of interest. The source worker VM does not appear to be having CPU or disk issues. We did notice higher than average memory usage during the period when the issue occurred but according to metrics there was still at least 4GB free at this point. In at least one case, the volume that was being streamed was itself a streamed volume that had been marked as a cache

Around the time of the issue we see a lot of `resource-cache-volume-disappeared-from-worker` errors as well as some `pq: insert or update on table "volumes" violates foreign key constraint "volumes_parent_id_fkey"` errors in the ATC logs.

We cannot reliably reproduce the problem on-demand right now but it happens consistently at least once a day on a mildly busy cluster

## Help
We are pretty stuck on what is going on. The errors themselves on the worker seem to stem from `os.Stat` either not finding the folder or it not being a directory which would suggest it has been deleted on disk but Concourse doesn't seem to think it was the thing that deleted the volume. Any suggestions about what to look at next are appreciated. We have debug logs for the workers and web nodes turned on for now.

**Comment by andy-paine** [comment]
I wonder if @evanchaoli has any thoughts since you are the expert on volume caching/streaming these days 😁

**Reply by andy-paine** [reply]
Is this just us observing the behaviour from this test? https://github.com/concourse/concourse/blob/master/integration/worker/streamed_resource_cache_test.go. I'm not entirely sure I understand why the cache on `worker2` needs GC-ing when `worker1` is pruned?

**Comment by evanchaoli** [comment]
@andy-paine I am pretty busy today, so I couldn't spend more time on this issue. As you have found the test case, I think you can easily debug with the test case via docker-compose.

When you debug the issue, you may disable all collectors by updating table `components`. Before prune worker1, check the following tables: `resource_caches`, `resource_cache_uses`, `worker_resource_caches`, `volumes`, then prune worker1, and see data changes in those table, then you may get some clue.

EDIT: I suddenly recalled why the test was designed. In `worker_resource_caches` table, there is a column `worker_base_resource_type_id` which points to origin worker base resource type id when the resource cache is created. After a volume is streamed to worker2, `worker_resource_caches` will container two tuples, both have the same `worker_base_resource_type_id` of worker1. So when worker1 is pruned, `worker_base_resource_types` is deleted due to `on cascade`, and delete cascade to `worker_resource_caches`. That's reason why worker2 will lose the resource cache after worker1 is pruned.

I think we should convert this topic to an issue. Because when a worker is pruned, if it has some volumes streamed to other workers, and some builds are running and using the resource cache, though the volumes exist on other workers, running builds may still fail.

To fix the issue, delete on `worker_base_resource_types` should not cascade to `worker_resource_caches`. Instead, if should only mark `worker_resource_caches` as invalid, and only GC those worker resource cache when there is no build using the resource cache.

**Reply by andy-paine** [reply]
I can replicate that locally and that seems to agree with what is happening in our real deployments:
* Resource is pulled onto worker A
* Resource is streamed to worker B and marked as cache
* Worker A lands
* Resource attempts to stream from worker B -> worker C as worker B has valid cache
* Resource has since been GC'd from worker B due to landing worker A

This is highly likely to happen when we have landing workers as we also use `limit-active-tasks` as effectively a task queue so the time that the `ArtifactRepository()` is populated with volumes that relate to caches and the time that the volume actually gets streamed from a worker can be quite large - easily long enough for the GC to sweep the volume away.

I think we may have an issue with not all logs making it off the machines as in subsequent occurrences of the bug I _have_ seen the volume try to be destroyed (destruction failed but I think the rename from `/live/<handle>` to `/dead/<handle>` went through so it could not longer be found for future streams).

Thanks @evanchaoli for your help (and your continued good work on this project!) - I think the integration test behaviour is exactly what we are seeing. Now we know exactly what is causing the problem is I agree it can be turned into a full issue which I'll write up.

**Comment by marco-m** [comment]
@odormond see the Background section. We could try to do the same...

**Reply by odormond** [reply]
@marco-m, we already operate with cache streamed volumes on.

The description matches very well with what we're seeing about the volume not found errors too. Also evancholi explanation totally makes sense to me.
